collect_app/src/androidTest/java/org/odk/collect/android/formentry/DefaultImageAnswerTest.java

```diff
@@ -0,0 +1,47 @@
+package org.odk.collect.android.formentry;
+
+import android.Manifest;
+
+import androidx.test.espresso.intent.rule.IntentsTestRule;
+import androidx.test.rule.GrantPermissionRule;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+import org.odk.collect.android.activities.FormEntryActivity;
+import org.odk.collect.android.support.CopyFormRule;
+import org.odk.collect.android.support.ResetStateRule;
+import org.odk.collect.android.support.pages.FormEntryPage;
+import org.odk.collect.android.test.FormLoadingUtils;
+
+import java.util.Collections;
+
+public class DefaultImageAnswerTest {
```

Comment:
I think we could drive these changes out in tests for the `Widget`s (Robolectric most likely) themselves rather than at an Espresso level. This would also let us test that the correct default image is loaded, as well as testing the `ImageView` appears. What do you think?

Reply:
I was thinking about it. We could add a test for every single widget like I did for [readOnly](https://github.com/opendatakit/collect/pull/3551) option. The problem is that we always get answers using` getAnswerText()` (from Javarosa) so we don't know if such an answer comes from a default value or it's just a previously saved answer. So in fact such a test (in every single widget) would be for saved answers not default values. The only way to make sure we are testing default values is using a form.

Reply:
I'm not meaning that we test the whole of the default value behaviour - like you say that's something that JavaRosa is partly responsible for. I mean that we could have a test for driving out that the image is shown (in `BaseImageWidget`) if there is a default (`getAnswerText` returns an image reference). In pseudocode I'm thinking:

1. Create temp image file
2. Set up mock reference to default image (`Helpers.createMockReference`)
3. Set up prompt with reference returned from `getAnswerText()` (`MockFormEntryPromptBuilder`)
4. Create `BaseImageWidget` with prompt
5. Check the widget's image view has loaded the bitmap - I think you can do this using Robolectric's `ShadowBitmapDrawable` (`shadowOf((BitmapDrawable) imageView.getDrawable()).getPath()`)

As far as I can see this would drive out the changes without needing Espresso. I might be misunderstanding something though.

Reply:
I believe that @grzesiek2010 is correct that what you've described, @seadowg, would not specifically test having an image default but would just check that an image is loaded when there's an image in the current form instance. I think it's a worthwhile test but it's not testing the new behavior here. In other words, I would expect it to pass on `master`.

The new behavior is doing something with a `jr://images/` URI and I think step 2 of what @seadowg has described wipes that out.

@seadowg's test would need to be augmented with some tests that fail today. I think that ideally we could add a component to the test that lets us verify that either a relative (local to the submission folder) URI or an absolute (`jr://images/` which means relative to the form definition folder) URI can be resolved and passed on to the image display machinery.

Also, I think the implementation is incorrect because it copies an image with a `jr://images/` URI to the submission folder which it shouldn't do.

Reply:
Hmmm I must be misunderstanding how it shiould work. My thinking was that kind of test would drive out the code in `getDefaultFilPath` that deals with `jr://images/`. Either way I think agree with how you think it should go!

Reply:
Well, wait a minute, maybe I'm the one not getting it. How are you thinking a relative URI and `jr://images/` URI would be differentiated in what you described?

Reply:
I think what I'd missed is that it did different things on whether the URI started with `jr://images/ ` rather than only setting `defaultFilePath` if the URI started with `jr://images/ `. I was seeing an `if` rather than an `if-else`. I'm guessing you're saying we need two tests (one for each kind of URI) to drive out the new behaviour?

Reply:
> I'm guessing you're saying we need two tests (one for each kind of URI) to drive out the new behaviour?

Yes. And on `master`, the test using a `jr://images/` URI should fail.

I added some comments around the implementation which I don't think is quite right. There really is no "default" concept at this point, we can't tell the difference between a value from a form definition vs a submission that's being re-entered.

Reply:
I added some unite tests for each image widget but still I think it would be good to keep this espresso tests as well.

Reply:
I put together an example of how I **think** we could drive out the changes to the `ImageWidget`s using just Robolectric. It requires an update to our `createMockReference` method so it can use real files.

https://gist.github.com/seadowg/a9dd682666b2ec7b8f9d242f6553e9ea

I'm very interested in seeing how far we can go with testing Widgets in isolation as I think it would let us write a lot fewer Espresso tests. Can totally see that it might just feel easier/better to stick with the Espresso test here though.

Reply:
@seadowg, when you say "drive out", you mean that in a test-driven development sense, right? As in write a test that describes the desired behavior at a high level which initially fails and use that test to get to a good code design? I just want to make sure that we're all on the same page. I do think that's a powerful approach that would be good to use more as opposed to writing some code and then fitting tests around it.

I think that both the tests that @grzesiek2010 added and the one @seadowg is proposing basically check to make sure that the `ReferenceManager` gets used to resolve a `jr://images` URI. @grzesiek2010's very much rely on the current structure of the code because the unit they test is the method that goes from an answer value in an instance to an actual path. What I like about @seadowg's in contrast is that it is at a higher level and verifies what the user ultimately cares about -- that the image that will be shown is the one they expect. It requires exposing a getter for the `imageView` but that change seems acceptable to me. The test remains high-level enough that it could support a refactor of the functionality it tests. @grzesiek2010, would a test like that make you feel comfortable dropping the Espresso test?

As I mentioned before, ideally we'd have an analogous test for an image without a `jr://images` prefix. I believe that would take a non-trivial refactor because of the dependency that `addCurrentImageToLayout` has on `FormController`.

Reply:
> when you say "drive out", you mean that in a test-driven development sense, right? As in write a test that describes the desired behavior which initially fails and use that test to get to a good code design?

Yes exactly. I find a nice trick for writing these after the fact is to comment out the changes in the code, write the tests for the behaviour, verify they they're red/fail and then bring the code back (uncomment) until you're green again.

> It requires exposing a getter for the imageView but that change seems acceptable to me.

Yeah this can seem weird given you'd usually avoid making something public just to test it - you want to test the "interface" and so as you were pointing out elsewhere it can be a sign that you need to change your design. In this case the `imageView` is "visible" to the user so actually is part of the "interface".

Reply:
> would a test like that make you feel comfortable dropping the Espresso test?

yes I think so but if it's only about test I would vote for testing it and merging as is and then working on tests in a separate pr.

Reply:
Works for me. Maybe @seadowg can take the test he's written, polish it up and send it in, also removing the Espresso test. Perhaps it will also be a reasonable time to pull out the `FormController` dependency and do the test for non-`jr://images` files?

Reply:
Sounds good to me!